find-static.cpp

```diff
@@ -54,42 +56,100 @@ int main(int argc, char **argv) {
     }
   }
 
-  for(const auto& module : loadedModules) {
-    for(auto const &function : module->functions()) {
-      if(function.isDeclaration()) {
+  for (const auto &module : loadedModules) {
+    for (auto const &function : module->functions()) {
+      if (function.isDeclaration()) {
         definedFunctions.erase(function.getName().str());
       }
     }
   }
-  
-  for(const auto &[key, value] : definedFunctions) {
+
+  for (auto &[key, value] : definedFunctions) {
     auto linkage = linkageTypes[value->getLinkage()];
-    if(linkage.length() == 0) {
+    if (linkage.length() == 0) {
       linkage = std::to_string(value->getLinkage());
     }
     auto visibility = visibilityTypes[value->getVisibility()];
+    
+    std::stringstream out;
+
+    // TODO: Const washing is a bad thing.
+    llvm::Module *module = (llvm::Module *)value->getParent();
+    auto callGraph = new llvm::CallGraph(*module);
+    bool viable = true;
 
-    if(value->getLinkage() != llvm::GlobalValue::InternalLinkage) {
+    for (const auto &[_, node] : *callGraph->getOrInsertFunction(value)) {
+      if (node->getFunction() == nullptr) {
+        continue;
+      }
+      const auto child = node->getFunction();
+      if (child->getLinkage() == llvm::GlobalValue::InternalLinkage) {
+        viable = false;
+      }
+      out << " - " << child->getName().str() << ": "
+          << visibilityTypes[child->getVisibility()] << " "
+          << linkageTypes[child->getLinkage()] << std::endl;
+    }
+
+    for (const llvm::User *user : value->users()) {
+      if (auto *instruction = llvm::dyn_cast<llvm::Instruction>(user)) {
+        auto function = instruction->getParent()->getParent();
+        if (function->getLinkage() == llvm::GlobalValue::InternalLinkage ||
+            instruction->getOpcode() == llvm::Instruction::Store) {
+          viable = false;
+        }
+        out << "Inst: " << instruction->getOpcodeName() << "<"
+            << instruction->getOpcode() << ">"
+            << "\t" << function->getName().str() << ": "
+            << visibilityTypes[function->getVisibility()] << " "
+            << linkageTypes[function->getLinkage()] << "\n";
+
+        for (const auto &[_, node] :
+             *callGraph->getOrInsertFunction(function)) {
+          if (node->getFunction() == nullptr) {
+            continue;
+          }
+          const auto child = node->getFunction();
+          if (child->getLinkage() == llvm::GlobalValue::InternalLinkage) {
+            viable = false;
+          }
+          out << " - \t\t" << child->getName().str() << ": "
+              << visibilityTypes[child->getVisibility()] << " "
+              << linkageTypes[child->getLinkage()] << std::endl;
+        }
+      } else if (auto *globalValue = llvm::dyn_cast<llvm::GlobalValue>(user)) {
+        out << "\t" << globalValue->getName().str() << "\n";
+        viable = false;
+      } else {
+        viable = false;
+        out << "\t"
```

Comment:
Not that I feel strongly about this, but would this be clearer as `out << "\tCONST\n";` on one line?

Reply:
alright I will change it like that.